cmake/org.eclipse.cdt.cmake.ui/src/org/eclipse/cdt/cmake/ui/internal/CMakePreferencePage.java

```diff
@@ -205,7 +367,108 @@ public boolean performOk() {
 		filesToAdd.clear();
 		filesToRemove.clear();
 
+		// Update Preferences for cmakeSupplier
+		try {
+			getPreferences().clear();
+			getPreferences().node(CMAKE_GENERATOR_LOCATION).clear();
+			getPreferences().putBoolean(ENABLE_USE_CMAKE_LOCATION, useCmakeToolLocation);
+			if (!cmakeLocation.isEmpty()) {
+				getPreferences().put(CMAKE_LOCATION, cmakeLocation);
+			}
+			int index;
+			for (index = 0; index < generatorLocations.length; index++) {
+				getPreferences().node(CMAKE_GENERATOR_LOCATION).put(String.format("location.%d", index), //$NON-NLS-1$
+						generatorLocations[index]);
+			}
+			getPreferences().flush();
+		} catch (BackingStoreException e) {
+			Activator.log(e);
+		}
 		return true;
 	}
 
+	private void updateCmakeToolGroupData() {
+		try {
+			useCmakeToolLocation = getPreferences().getBoolean(ENABLE_USE_CMAKE_LOCATION, false);
+			cmakeLocation = getPreferences().get(CMAKE_LOCATION, EMPTY_STRING);
+			List<String> genlocs = new ArrayList<>();
+			for (String key : getPreferences().node(CMAKE_GENERATOR_LOCATION).keys()) {
+				genlocs.add(getPreferences().node(CMAKE_GENERATOR_LOCATION).get(key, EMPTY_STRING));
+			}
+			generatorLocations = genlocs.toArray(new String[0]);
+		} catch (BackingStoreException e) {
+			Activator.log(e);
+		}
+	}
+
+	private String resolveVariableValue(String value) {
+		try {
+			ICdtVariableManager vm = CCorePlugin.getDefault().getCdtVariableManager();
+			return vm.resolveValue(value, null, EMPTY_STRING, null);
+		} catch (CdtVariableException e) {
+			Activator.log(e);
+		}
+		return null;
+	}
+
+	private String getVariableDialog(Shell shell, ICConfigurationDescription cfgd) {
+		ICdtVariableManager vm = CCorePlugin.getDefault().getCdtVariableManager();
+		BuildVarListDialog dialog = new BuildVarListDialog(shell, vm.getVariables(cfgd));
+		dialog.setTitle(Messages.VariablesDialog_Title);
+		if (dialog.open() == Window.OK) {
+			Object[] selected = dialog.getResult();
+			if (selected.length > 0) {
+				String s = ((ICdtVariable) selected[0]).getName();
+				return "${" + s.trim() + "}"; //$NON-NLS-1$//$NON-NLS-2$
+			}
+		}
+		return null;
+	}
+
+	private class EditGeneratorLocationDialog extends Dialog {
+
+		private String fTitle;
+		private FileListControl fListEditor;
+		private String[] fGeneratorLocations;
+
+		public EditGeneratorLocationDialog(Shell parentShell, String title, String[] generatorLocations) {
+			super(parentShell);
+			this.fGeneratorLocations = generatorLocations;
+		}
+
+		@Override
+		protected void configureShell(Shell shell) {
+			super.configureShell(shell);
+			if (fTitle != null) {
+				shell.setText(fTitle);
+			}
+		}
+
+		@Override
+		protected Control createDialogArea(Composite parent) {
```

Comment:
In the Generator locations(s) field, after clicking Edit..., the Generator location list dialog appears and when the user clicks Add... button the Enter Value dialog appears, shown in the screenshot below:

![image](https://github.com/user-attachments/assets/54ca1024-77da-4034-9af0-69132dc2fcca)

Is it possible to add a directory browse button to help the user add the path?

Reply:
Thanks @DangMinhTam382 for providing a screenshot.
I think that's fine
😀